concent_api/core/views.py

```diff
@@ -1220,6 +1225,632 @@ def store_message_and_message_status(
+            assert subtask.state is not Subtask.SubtaskState.ADDITIONAL_VERIFICATION.name
+
+        logging.log_changes_in_subtask_states(
+            b64encode(client_public_key),
+            clients_subtask_list.count(),
+        )
+    else:
+        logging.log_no_changes_in_subtask_states(
+            b64encode(client_public_key)
+        )
+
+
+def update_subtask_state(
+    subtask,
+    state,
+):
+    logging.log_change_subtask_state_name(
+        subtask.state,
+        state,
+    )
+    subtask.state    = state
+    subtask.next_deadline = None
+    subtask.full_clean()
+    subtask.save()
+
+
+def store_pending_message(
+    response_type       = None,
+    client_public_key   = None,
+    queue               = None,
+    subtask             = None,
+    payment_message     = None,
+):
+    client          = Client.objects.get_or_create_full_clean(client_public_key)
+    receive_queue   = PendingResponse(
+        response_type   = response_type.name,
+        client          = client,
+        queue           = queue.name,
+        subtask         = subtask,
+    )
+    receive_queue.full_clean()
+    receive_queue.save()
+    if payment_message is not None:
+        payment_committed_message = PaymentInfo(
+            payment_ts                  = datetime.datetime.fromtimestamp(payment_message.payment_ts, timezone.utc),
+            task_owner_key_bytes        = decode_key(payment_message.task_owner_key),
+            provider_eth_account_bytes  = payment_message.provider_eth_account,
+            amount_paid                 = payment_message.amount_paid,
+            recipient_type              = payment_message.recipient_type.name,  # pylint: disable=no-member
+            amount_pending              = payment_message.amount_pending,
+            pending_response            = receive_queue
+        )
+        payment_committed_message.full_clean()
+        payment_committed_message.save()
+        subtask_id = None
+    else:
+        subtask_id = subtask.subtask_id
+
+    logging.log_new_pending_response(
+        response_type.name,
+        queue.name,
+        subtask_id,
+        client.public_key,
+    )
+
+
+def handle_messages_from_database(
+    client_public_key:  bytes                   = None,
+    response_type:      PendingResponse.Queue   = None,
+):
+    assert client_public_key    not in ['', None]
+
+    database_message = PendingResponse.objects.filter(
+        client__public_key = b64encode(client_public_key),
+        queue              = response_type.name,
+        delivered          = False,
+    ).order_by('created_at').first()
+
+    if database_message is None:
+        return None
+
+    assert database_message.response_type_enum in set(PendingResponse.ResponseType)
+
+    if database_message.response_type != PendingResponse.ResponseType.ForcePaymentCommitted.name:
+        assert database_message.subtask.state_enum in set(Subtask.SubtaskState) - {Subtask.SubtaskState.ADDITIONAL_VERIFICATION}
```

Comment:
This assumption is incorrect. You're basically saying that if the receive queue contains any message other than `ForcePaymentCommitted`, the task must not currently be in the `ADDITIONAL_VERIFICATION` state. But remember that the client can call `receive/` at any time, even long after we have gone through several states and finished the task. So you can still have the old `ForceReportComputedTaskResponse` in queue (that was not delivered yet) while the task is already in `ADDITIONAL_VERIFICATION`.

Please review this method and make sure you don't check the current task state here. Nothing here should depend on the state.

Reply:
Deleted state checking in views.